compiler/src/dotty/tools/dotc/transform/ArrayApply.scala

```diff
@@ -22,27 +19,71 @@ class ArrayApply extends MiniPhase {
 
   override def description: String = ArrayApply.description
 
-  override def transformApply(tree: tpd.Apply)(using Context): tpd.Tree =
+  private val TransformListApplyBudgetKey = new Property.Key[Int]
+  private def transformListApplyBudget(using Context) = ctx.property(TransformListApplyBudgetKey).getOrElse(8)
+
+  override def prepareForApply(tree: Apply)(using Context): Context =
+    if isSeqApply(tree) then
+      val args = seqApplyArgsOrNull(tree)
+      if args != null then
+        ctx.fresh.setProperty(TransformListApplyBudgetKey, transformListApplyBudget - args.elems.length)
+      else ctx
+    else ctx
+
+  override def transformApply(tree: Apply)(using Context): Tree =
     if isArrayModuleApply(tree.symbol) then
-      tree.args match {
-        case StripAscription(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: ct :: Nil
+      tree.args match
+        case StripAscription(Apply(wrapRefArrayMeth, (seqLit: JavaSeqLiteral) :: Nil)) :: ct :: Nil
             if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) && elideClassTag(ct) =>
           seqLit
 
-        case elem0 :: StripAscription(Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil)) :: Nil
+        case elem0 :: StripAscription(Apply(wrapRefArrayMeth, (seqLit: JavaSeqLiteral) :: Nil)) :: Nil
             if defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) =>
-          tpd.JavaSeqLiteral(elem0 :: seqLit.elems, seqLit.elemtpt)
+          JavaSeqLiteral(elem0 :: seqLit.elems, seqLit.elemtpt)
 
         case _ =>
           tree
-      }
+
+    else if isSeqApply(tree) then
+      val args = seqApplyArgsOrNull(tree)
+      if args != null && (transformListApplyBudget > 0 || args.elems.isEmpty) then
+        val consed = args.elems.foldRight(ref(defn.NilModule)): (elem, acc) =>
+          New(defn.ConsType, List(elem.ensureConforms(defn.ObjectType), acc))
+        consed.cast(tree.tpe)
+      else tree
 
     else tree
 
   private def isArrayModuleApply(sym: Symbol)(using Context): Boolean =
     sym.name == nme.apply
     && (sym.owner == defn.ArrayModuleClass || (sym.owner == defn.IArrayModuleClass && !sym.is(Extension)))
 
+  private def isListApply(tree: Tree)(using Context): Boolean =
+    (tree.symbol == defn.ListModule_apply || tree.symbol.name == nme.apply) && appliedCore(tree).match
+      case Select(qual, _) =>
+        val sym = qual.symbol
+        sym == defn.ListModule
+        || sym == defn.ListModuleAlias
+      case _ => false
+
+  private def isSeqApply(tree: Tree)(using Context): Boolean =
+    isListApply(tree) || tree.symbol == defn.SeqModule_apply && appliedCore(tree).match
+      case Select(qual, _) =>
+        val sym = qual.symbol
+        sym == defn.SeqModule
+        || sym == defn.SeqModuleAlias
+        || sym == defn.CollectionSeqType.symbol.companionModule
+      case _ => false
+
+  private def seqApplyArgsOrNull(tree: Apply)(using Context): JavaSeqLiteral | Null =
```

Comment:
We usually avoid nulls like that one. Maybe define a `SeqApply(args)` extractor that covers both `isSeqApply` and `seqApplyArgsOrNull`.

Example use:
```diff
   override def prepareForApply(tree: Apply)(using Context): Context =
-    if isSeqApply(tree) then
-      val args = seqApplyArgsOrNull(tree)
-      if args != null then
-        ctx.fresh.setProperty(TransformListApplyBudgetKey, transformListApplyBudget - args.elems.length)
-     else ctx
-    else ctx
+    tree match
+      case SeqApply(args) => 
+        ctx.fresh.setProperty(TransformListApplyBudgetKey, transformListApplyBudget - args.elems.length)
+      case _ => 
+        ctx
```